File: 2016/8/solution.js
const fs = require('fs');
const path = require('path');

const input = fs.readFileSync(path.join(__dirname, 'input.txt'), 'utf-8', (err, data) => {
  if (err) {
    console.log(err)
  } else {
    return data;
  }
}).split('\n');

const screen = new Array(6).fill(null).map((a) => new Array(50).fill(' '));

const re = /\d+/g;

for (const line of input) {
  const nums = line.match(re).map(Number);
  if (line.includes('rect')) {
    for (let j = 0; j < nums[0]; j++) {
      for (let i = 0; i < nums[1]; i++) {
        screen[i][j] = '#';
      }
    }
  } else if (line.includes('row')) {
    screen[nums[0]] = screen[nums[0]].slice(50 - nums[1]).concat(screen[nums[0]].slice(0, 50 - nums[1]))
  } else {
    const column = screen.map((a) => a[nums[0]]);
    for (let i = nums[1]; i < 6; i++) {
      screen[i][nums[0]] = column[i - nums[1]];
    }
    for (let i = 0; i < nums[1]; i++) {
      screen[i][nums[0]] = column[i + 6 - nums[1]];
    }
  }
}

let count = 0;
for (const row of screen) {
  count += row.reduce((a, b) => a + (b === '#' ? 1 : 0), 0);
  console.log(row.join(''));
}

console.log(count);

File: 2016/8/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := strings.Split(string(data), "\n")
	screen := makeScreen()
	fmt.Println(part1(puzzleInput, screen))
	part2(screen)
}

const (
	ScreenHeight = 6
	ScreenWidth  = 50
)

func rect(screen *[][]bool, n1, n2 int) {
	for i := range n1 {
		for j := range n2 {
			(*screen)[j][i] = true
		}
	}
}

func rotateRow(screen *[][]bool, n1, n2 int) {
	(*screen)[n1] = append((*screen)[n1][ScreenWidth-n2:], (*screen)[n1][:ScreenWidth-n2]...)

}

func rotateColumn(screen *[][]bool, n1, n2 int) {
	tmp := make([]bool, ScreenHeight)
	for row := range ScreenHeight {
		tmp[(row+n2)%ScreenHeight] = (*screen)[row][n1]
	}
	for row := range ScreenHeight {
		(*screen)[row][n1] = tmp[row]
	}
}

func makeScreen() *[][]bool {
	screen := make([][]bool, ScreenHeight)
	for row := range ScreenHeight {
		screen[row] = make([]bool, ScreenWidth)
	}
	return &screen
}

func printScreen(screen *[][]bool) {
	var sb strings.Builder
	for _, row := range *screen {
		sb.Reset()
		for _, ch := range row {
			if ch {
				sb.WriteRune('█')
			} else {
				sb.WriteRune(' ')
			}
		}
		fmt.Println(sb.String())
	}
}

func part1(puzzleInput []string, screen *[][]bool) int {
	instructionRegex := regexp.MustCompile(`^([a-z ]+)\D+(\d+)\D+(\d+)$`)
	for _, line := range puzzleInput {
		match := instructionRegex.FindStringSubmatch(line)
		if match == nil {
			panic(fmt.Sprintf("Error matching instruction regex with : %s", line))
		}
		n1, err := strconv.Atoi(match[2])
		if err != nil {
			panic(err)
		}
		n2, err := strconv.Atoi(match[3])
		if err != nil {
			panic(err)
		}
		switch match[1] {
		case "rect":
			rect(screen, n1, n2)
		case "rotate row y":
			rotateRow(screen, n1, n2)
		case "rotate column x":
			rotateColumn(screen, n1, n2)
		default:
			panic(fmt.Sprintf("Unrecognized instruction: %s ", match[1]))
		}
	}
	res := 0
	for _, row := range *screen {
		for _, p := range row {
			if p {
				res++
			}
		}
	}
	return res
}

func part2(screen *[][]bool) {
	printScreen(screen)
}
